src/core/mavlink_commands.cpp

```diff
@@ -278,39 +275,64 @@ void MavlinkCommandSender::receive_timeout()
+
+    if (!work) {
+        // Nothing to do.
+        return;
+    }
+
+    if (work->already_sent) {
+        return;
+    }
+
+    LogDebug() << "sending it the first time (" << work->mavlink_command << ")";
+    work->time_started = _parent.get_time().steady_time();
 
     {
-        LockedQueue<Work>::Guard work_queue_guard(_work_queue);
-        auto work = work_queue_guard.get_front();
+        std::lock_guard<std::mutex> lock(_sent_commands_mutex);
+        auto was_inserted =
+            _sent_commands.insert(std::make_pair(work->mavlink_command, work)).second;
+
+        // The command is inserted in the list only if another command with the same id does
+        // not exist. If such an element exists, we refuse to send the new command.
+        if (!was_inserted) {
+            LogWarn()
+                << "A command cannot be sent if another command with the same command_id is still processing! Ignoring command "
+                << static_cast<int>(work->mavlink_command);
+            auto temp_callback = work->callback;
+            auto temp_result = std::make_pair<Result, float>(Result::CommandDenied, NAN);
+
+            if (temp_callback != nullptr) {
+                call_callback(temp_callback, temp_result.first, temp_result.second);
+            }
 
-        if (!work) {
-            // Nothing to do.
             return;
         }
 
-        if (!work->already_sent) {
-            // LogDebug() << "sending it the first time (" << work->mavlink_command << ")";
-            work->time_started = _parent.get_time().steady_time();
-            if (!_parent.send_message(work->mavlink_message)) {
-                LogErr() << "connection send error (" << work->mavlink_command << ")";
-                temp_callback = work->callback;
-                temp_result = {Result::ConnectionError, NAN};
-                work_queue_guard.pop_front();
-            } else {
-                work->already_sent = true;
-                _parent.register_timeout_handler(
-                    std::bind(&MavlinkCommandSender::receive_timeout, this),
-                    work->timeout_s,
-                    &_timeout_cookie);
+        if (!_parent.send_message(work->mavlink_message)) {
+            LogErr() << "connection send error (" << work->mavlink_command << ")";
+            auto temp_callback = work->callback;
+            auto temp_result = std::make_pair<Result, float>(Result::ConnectionError, NAN);
+
+            if (temp_callback != nullptr) {
+                call_callback(temp_callback, temp_result.first, temp_result.second);
             }
+
+            _sent_commands.erase(work->mavlink_command);
+
+            return;
         }
     }
 
-    if (temp_callback != nullptr) {
-        call_callback(temp_callback, temp_result.first, temp_result.second);
-    }
+    work->already_sent = true;
+    _parent.register_timeout_handler(
+        std::bind(&MavlinkCommandSender::receive_timeout, this, work->mavlink_command),
```

Comment:
I usually try to go to lambda instead of `std::bind`.

Reply:
Agreed, but here I actually just moved around the previous code (which is admittedly not reflected well in the git diff), I did not want to refactor too much in this PR (if that's okay for you).

Reply:
Yap, that's fine.